### PR TITLE
[FLINK-3030][web frontend] Enhance dashboard to show execution attempts

### PR DESCRIPTION
**Pull request by mushketyk**
I've added support for displaying execution attempts in the dashboard. 

The JIRA item also includes that we should add accumulators for different attempts, but it seems that accumulators are going to be deprecated (judging from this JIRA item https://issues.apache.org/jira/browse/FLINK-4527) so I didn't implement that.

Here are few screenshots of the new feature:
A list of opened attempts are collapsed:

attempts collapsed

A list of opened attempts is opened:

opened attempts

If there is no attempts the button is not displayed:

no attempts
- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)
- [x] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added
- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/HandlersUtils.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+import org.apache.flink.api.common.accumulators.LongCounter;
+import org.apache.flink.runtime.accumulators.AccumulatorRegistry;
+import org.apache.flink.runtime.execution.ExecutionState;
+import org.apache.flink.runtime.executiongraph.Execution;
+import org.apache.flink.runtime.instance.InstanceConnectionInfo;
+
+import java.io.IOException;
+import java.util.Map;
+
+public final class HandlersUtils {
```
Comment (StephanEwen) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/HandlersUtils.java]: How about giving this a more revealing name. Like `JsonGenerators`?
You could also simply move the `serializeExecutionAttempt(...)` method as a static method to the `SubtaskExecutionAttemptDetailsHandler`.

File: flink-runtime-web/web-dashboard/web/css/index.css [hunk 2]
```diff
@@ -228,7 +228,6 @@
   overflow: hidden;
   padding: 2px 4px;
   font-size: 14px;
-  -webkit-border-radius: 2px;
```
Comment (StephanEwen) [inline flink-runtime-web/web-dashboard/web/css/index.css]: How did this come to be? Are you using a different version of glup or the dependencies to build this?

**Comment by StephanEwen** [comment]
Good change, all in all, good work!

Few minor comments, otherwise +1

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/HandlersUtils.java (hunk 1 above)
Comment (mushketyk) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/HandlersUtils.java]: Good point. I would prefer to have a separate class `JsonGenerators` that will have just one responsibility.

File: flink-runtime-web/web-dashboard/web/css/index.css (hunk 2 above)
Comment (mushketyk) [inline flink-runtime-web/web-dashboard/web/css/index.css]: Good catch. I'll take a look at what `gulp` version am I using and check how I can fix it.
Could you please check what version of `gulp` are you using? I don't think it is documented anywhere.

**Comment by mushketyk** [comment]
Thank you for the review @StephanEwen
I'll update the PR in the evening.

**Comment by mushketyk** [comment]
@StephanEwen I've renamed the class as you suggested.
Regarding the CSS change, I don't know why these changes are added. I have the required (according to README.md) `gulp` version:
```
> $ gulp -version 
[21:06:35] CLI version 3.9.1
[21:06:35] Local version 3.9.1

```

I wonder if a `.styl` file was changed in one of the previous commits but it was not compiled to `.css`.

**Comment by mushketyk** [comment]
@StephanEwen Should I change anything else in this PR?

**Comment by mushketyk** [comment]
Is there anything else that should be updated in this PR? It has been in review for more than a month.

**Comment by StephanEwen** [comment]
Good to go from my side.
@rmetzger and @iampeter - since you are working on the web UI currently, can you try and merge this?

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/JsonGenerators.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+		LongCounter readBytes;
+		LongCounter writeBytes;
+		LongCounter readRecords;
+		LongCounter writeRecords;
+
+		if (metrics != null) {
+			readBytes = (LongCounter) metrics.get(AccumulatorRegistry.Metric.NUM_BYTES_IN);
+			writeBytes = (LongCounter) metrics.get(AccumulatorRegistry.Metric.NUM_BYTES_OUT);
+			readRecords = (LongCounter) metrics.get(AccumulatorRegistry.Metric.NUM_RECORDS_IN);
+			writeRecords = (LongCounter) metrics.get(AccumulatorRegistry.Metric.NUM_RECORDS_OUT);
```
Comment (zentol) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/JsonGenerators.java]: that is correct. It also has to be rebased due to the introduction of the ArchivedExecutionGraph.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java [hunk 4]
```diff
@@ -241,6 +241,18 @@ public Execution getPriorExecutionAttempt(int attemptNumber) {
 		}
 	}
 
+	public Execution getExecutionAttempt(int attemptNumber) {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java]: i would prefer if you would instead use getCurrentExecutionAttempt if attemptNumber==0 and getPriorExecutionAttempt if attemptNumber>0, for the sole purpose of not having to many methods in the ExecutionVertex class.

**Comment by iampeter** [comment]
@rmetzger will you be merging? if so, `flink-runtime-web/web-dashboard/web/js/index.js` just needs a `gulp`

**Comment by rmetzger** [comment]
Okay, I'll run gulp when merging.

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/JsonGenerators.java (hunk 3 above)
Comment (mushketyk) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/JsonGenerators.java]: Ok. I'll rebase the change and fix merge conflicts.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java (hunk 4 above)
Comment (mushketyk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java]: Ok. Will do this.

**Comment by mushketyk** [comment]
@rmetzger @iampeter 
I've rebased the code and updated it according to your suggestions.

**Comment by uce** [comment]
Really like this! I will check it out and try to play around with it a little and merge if everything works as expected. How does this interplay with https://issues.apache.org/jira/browse/FLINK-5645 @zentol? The `SubtaskAllExecutionAttemptsHandler` uses `execAttempt.getIOMetrics()`.

---

In general, I think the web frontend requires a general overhaul making execution attempts and failures first class citizens. The current state of this won't scale well to long running deployments where potentially many failures happen over time. Then you have to scroll to each subtask and find the failed attempt you are looking for etc.

**Comment by uce** [comment]
It worked well, but I'm a little hesitant to merge it as it is. Running with higher parallelism than 1 I think that the button is pretty invasive:

screen shot 2017-01-27 at 11 50 37

Can we make this a little less eye-catching? Maybe just a link next to the attempt number.

**Comment by mushketyk** [comment]
Hi @uce 

Thank you for your review.
I'll update the UI in the next couple of days.

**Comment by mushketyk** [comment]
Hi @uce,

I've rebased changes and updated the UI according to your suggestion. Now it looks like:
new attempts link

What do you think about this?

**Comment by aljoscha** [comment]
I'm closing this as "Abandoned", since there is no more activity and the code base has moved on quite a bit. Please re-open this if you feel otherwise and work should continue.

There is a new dashboard implementation now.

**Pull request closed without merging**